### PR TITLE
feat(#564): ingest instructors from M365 CSV

### PR DESCRIPTION
**Pull request by Fybex**
## #564 — Ingest instructors from the Microsoft 365 user export

Replaces the unreliable scraper-derived instructor extraction with a deterministic ingest from the M365 user CSV, and simplifies the `Instructor` model.

### What changed
- **Scraper no longer writes `Instructor` / `CourseInstructor`.** Removed `InstructorExtractor`, the M2M injector path, and the academic-degree/title enrichment heuristics. The raw `teachers` free-text string is *kept* (parser → parsed model) as a fallback signal for the multi-select work in #563.
- **`Instructor` model simplified** to `(first_name, patronymic, last_name, email)`. Dropped `academic_degree` / `academic_title` (could not be inferred reliably). `email` is `unique`, `db_index`.
- **New management command** `ingest_instructors_from_csv <path> [--dry-run]`:
  - **Internal-tenant filter** — `@ukma.edu.ua`, `userType == Member`, no `#ext#` guest marker.
  - **Service-account filter** — a `displayName` is kept only if it is structurally a personal name: at least two letter-only tokens (apostrophes and hyphens allowed), normal capitalisation, and no digits. A short residual `_SERVICE_WORDS` denylist catches the hard case structure cannot: title-cased English service mailboxes (e.g. `Backup Account`). The UPN local-part filter drops reserved aliases and underscore mailboxes.
  - **Students are intentionally kept.** They share the `@ukma.edu.ua` domain and cannot be told apart from teaching staff in the export, so they stay in the pool; the ranked `GET /instructors/` list (in #563) surfaces actually-rated teachers first, leaving students in the searchable tail.
  - Cyrillic (`Last First [Patronymic]`) vs Latin (`First Last`) name parsing.
  - `update_or_create` keyed on lowercased email → idempotent.
- **Migration `0028`** — schema-only (no data migration): drop the two academic columns, add unique `email`.
- Domain-model PUML + OpenAPI spec regenerated.

### Model naming
`Instructor` is the **broad candidate pool** sourced from the M365 directory, deliberately distinct from the scraped `Student` model (a person may appear in both). Most rows are only *candidate* instructors; a row becomes an *actual* instructor by being rated, and the ranked list surfaces those first. This intent is now documented on the model so the over-inclusive pool is not mistaken for "confirmed teaching staff". No rename — the name reflects the table's purpose (people creditable as an instructor), parallel to `Student`.

### Filter validation against the real export
The filter rules were validated against an actual M365 export (~19.6k rows):
- The Ukrainian apostrophe ships as a **grave accent (U+0060)** or **left single quote (U+2018)**, not the straight `'`. The token check now accepts these variants — without it ~**456 real people** (Дар`я, Мар`яна, В`ячеслав, …) were misclassified as service accounts and silently dropped.
- Recognising names structurally (digit-reject + every-token-letters) replaced the hand-curated forbidden-character set and additionally caught Cyrillic org/room accounts the English-only word list missed (e.g. lab rooms ending in a number), with **zero** real-people loss.

### ⚠️ Deploy runbook (populated environments — staging/live)
`email` is `unique` with no default, so the `Instructor` table **must be empty** when `0028` runs.

1. **DB dump first.**
2. Manually purge the scraper-junk instructor rows in the DB (cascades clear the join). This is a deliberate, dumped-first step — the command has no `--purge` flag. Free-text `Rating.instructor` on existing ratings is **untouched**.
3. `migrate rating_app 0028`
4. `ingest_instructors_from_csv <export.csv>` (purely additive / idempotent upsert)
5. Deploy webapp with the regenerated orval types.

> Note: on staging + live the junk `Instructor` rows were already purged ahead of this deploy, so the table is empty and `0028` applies cleanly. Use the **newest** export CSV.

### Tests
- Ingest command: name parsing, every filter branch, apostrophe/digit edge cases, encoding fallback (UTF-8/UTF-16), `--dry-run` no-op, idempotent re-run. Fixtures are synthetic (no PII).
- Full backend suite green.

### Dependency / CI
- Bumped `vitest` to `^4.1.0` and forced `esbuild >=0.28.1` (override) to clear a freshly-published critical/high advisory (`GHSA-5xrq-8626-4rwp`, `GHSA-gv7w-rqvm-qjhr`) that started failing `pnpm audit --audit-level high`. Same pins exist on `main`, so this is not branch-specific. Lint, 241 webapp tests, and the production build all pass on the upgrade.

### Notes
- Backfilling existing `Rating.instructor` free text → FK is **out of scope** (tracked as a later MR after #563).

## Summary by CodeRabbit

* **New Features**
  * Instructors now include email addresses in profiles.
  * Bulk CSV import for instructors with dry-run, encoding fallbacks, validation, and idempotent upserts.

* **Changes**
  * Academic degree and title fields removed from instructor data and APIs.
  * Email is a required, unique identifier for instructors.
  * Admin listings, serializers, and ingestion flows updated to use email.

### CODE REVIEW

File: src/backend/rating_app/management/commands/ingest_instructors_from_csv.py [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+    return tokens[0], "", " ".join(tokens[1:])
+
+
+def _open_csv(path: str):
+    for encoding in ("utf-8-sig", "utf-16"):
+        try:
+            return open(path, encoding=encoding, newline="")
+        except UnicodeError:
+            continue
+    raise CommandError(f"Unable to decode CSV at {path} as utf-8-sig or utf-16")
```
[the remark this answers is not in the document]
Comment (Fybex) [reply src/backend/rating_app/management/commands/ingest_instructors_from_csv.py:201]: Fixed in c20e231e. `open()` defers decoding, so the probe now does `read(512)` + `seek(0)` inside the `try` before returning the handle — the utf-16 fallback is reachable and an undecodable file raises a clean `CommandError`. Added tests for the UTF-16 path and the undecodable case.

[the remark this answers is not in the document]
Comment (Fybex) [reply src/backend/rating_app/management/commands/ingest_instructors_from_csv.py:201]: Fixed in c20e231e. `open()` defers decoding, so the probe now does `read(512)` + `seek(0)` inside the `try` before returning the handle — the utf-16 fallback is reachable and an undecodable file raises a clean `CommandError`. Added tests for the UTF-16 path and the undecodable case.

**Comment by Fybex** [comment]
@coderabbitai review — the encoding fallback finding was fixed in c20e231e (probe-read inside the try + tests for the UTF-16 and undecodable cases). Please re-review; the CHANGES_REQUESTED is stale.

**Review by stasiaaleks — APPROVED** [review]

**Review by katerynabratiuk — APPROVED** [review]
